### ISSUE TITLE
PAY request returns SSL connect error

### ISSUE THREAD

**Issue by dnich** [paypal/adaptivepayments-sdk-php#64]
Yesterday I was testing fine, then today, I cannot connect to the paypal sandbox server.  When I make a PAY request, it returns SSL connect error.  There were no changes to my code.

I found this post in the PHP SDK issues https://github.com/paypal/PayPal-PHP-SDK/issues/474  and wonder if it is related?  My server supports TLS 1.2 and OpenSSL 1.0.1e-fips and I am using v3.9.1 of the adaptivepayments SDK PHP

**Comment by kenjohnson** [comment]
I have the same problem as you on one of my hosts. 
On the other host it works okay. 
  
  host 1
     I have same problem as you. My client
	 has Apache config set like this:
	 SSLProtocol all -SSLv2
	 I have tried PayPal-PHP-SDK-1.6.2
	 and 1.6.4 and I cannot make payment
	 with either one of them.
	 If anyone can suggest a better
	 SSLProtocol setting let me know. 
	 
  host 2
    I am running in the cloud, Amazon
	Web Services (AWS). I am running
	a LAMP stack. I just set it up a 
    couple of weeks ago. I don't have the 
    problem there. I can make payments. 
    I am running PayPal-PHP-SDK-1.6.2 

  I looked at the link suggested by @pp-randy,
on #474
   and I ran 
   ```   openssl s_client -connect tlstest.paypal.com:443 ```
   on both hosts. 
   On host1 it failed, 

 On host 2 it succeed

**Comment by dnich** [comment]
FWIW 
openssl s_client -connect tlstest.paypal.com:443  connects over TLSv1.2 successfully from my dedicated server.

**Comment by kenjohnson** [comment]
It turns out that on our host1,  we have an older version of openssl that does 
not support TLSv1.1 or TLSv1.2.  It is interesting that you have openssl working,
and are still having the problem. There must be something else going on.

**Comment by kenjohnson** [comment]
I think I have an idea of what might be going on.  

If your openssl is not up to date the upgrade to 
the SDK will not help you.  

There is an excellent stack overflow answer here to
verify if Curl is using TLS.
http://stackoverflow.com/questions/27904854/verify-if-curl-is-using-tls

Here's a copy of the first example.
```php
<?php
$ch = curl_init('https://www.howsmyssl.com/a/check');
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
$data = curl_exec($ch);
curl_close($ch);

$json = json_decode($data);
echo $json->tls_version;
?>
```

If I run this on my host1, the host that is failing, I get,
        TLS 1.0
If I run it on my AWS host, I get,
        TLS 1.2

The next test, tests what openssl can do. 
Here is a copy of that test.  
``` php
 <?php
$curl_info = curl_version();
echo $curl_info['ssl_version'];
?>
```
If I run this on my host1, the host that is failing, I get,
    NSS/3.13.5.0
If I run it on my AWS host, I get,
    NSS/3.19.1 

In that post, they reference the release notes; 
you will see that NSS supports TLS v1.1 in 3.14
and supports TLS v1.2 in 3.15

So, on my host that is failing, I have 3.13, and
on the host that is succeeding I have 3.19; this
makes sense.  

Now, I think that's why the PayPal SDK fix is resulting
in intermittent results.
In that example above, the first one, if you added 
```php
    curl_setopt($ch, CURLOPT_SSLVERSION, CURL_SSLVERSION_TLSv1_2 );
```
it would help some people who had openssl >= 3.15, but those 
who had openssl less than that would be out of luck. 

This might explain why it is working for some people and not
other people.

**Comment by dnich** [comment]
When I run these to bits of php, I get unexpected results. 

TLS 1.0
NSS/3.16.2.3 Basic ECC

The first line seems related to this problem, but the second one implies that I should have 1.2 support (as of 3.15 according to the post).  By the way, I don’t see TLS version support listed in the NSS release notes, I am trusting the poster is correct.

So if NSS can support TLS v1.2 but Curl chooses not to........ Hmmm.  This could be part of the problem.  But now I don’t know why this is happening also.

**Comment by dnich** [comment]
The TLS 1.0 result from curl got me thinking that curl might be the problem.  A little more digging in that direction lead me to this:

http://stackoverflow.com/questions/26452755/php-curl-with-nss-is-probably-using-sslv3-insted-of-tls-when-connecting-to-htt

and the relevant quote:
> in the version history of curl you will find a bug with ECC ciphers and the NSS library (which you use) which is only fixed in curl version 7.36 "nss: allow to use ECC ciphers if NSS implements them". 

I updated curl to v7.46.0 (latest) and ran this test again.  Curl now reports TLS 1.2.  So I tried connecting to PayPal again and was successful!

Thanks for your help, kenjohnson.  Hope this solves it for you as well.
